cmd/initialize_with_genesis.cpp

```diff
@@ -0,0 +1,335 @@
+        !genesis_json.contains("config") || !genesis_json["config"].contains("chainId") ||
+        !genesis_json["config"]["chainId"].is_number()) {
+        SILKWORM_LOG(LogLevel::Error) << "Incomplete Genesis File" << std::endl;
+        return -1;
+    }
+
+    bool res{false};
+    try {
+        // Prime directories and DB
+        lmdb::DatabaseConfig db_config{out};
+        db_config.set_readonly(false);
+        db_config.map_size = map_size;
+        std::shared_ptr<lmdb::Environment> env{lmdb::get_env(db_config)};
+        std::unique_ptr<lmdb::Transaction> txn{env->begin_rw_transaction()};
+        db::table::create_all(*txn);
+
+        auto block_number{Bytes(8, '\0')};
+        evmc::bytes32 root_hash;
+
+        if (genesis_json.contains("alloc")) {
+            if (!genesis_json["alloc"].is_object()) {
+                throw std::invalid_argument("alloc member is not object");
+            }
+
+            // Filling account + constructing genesis root hash
+            std::map<evmc::bytes32, Bytes> account_rlp;
+            // Tables used
+            auto plainstate_table{txn->open(db::table::kPlainState)};
+            auto account_changeset_table{txn->open(db::table::kPlainAccountChangeSet)};
+
+            // Iterate over allocs
+            for (auto& item : genesis_json["alloc"].items()) {
+                if (!item.value().is_object() || !item.value().contains("balance") ||
+                    !item.value()["balance"].is_string()) {
+                    throw std::invalid_argument("alloc address " + item.key() + " has badly formatted allocation");
+                }
+
+                auto address_bytes{from_hex(item.key())};
+                if (address_bytes == std::nullopt || address_bytes.value().length() != kAddressLength) {
+                    throw std::invalid_argument("alloc address " + item.key() +
+                                                " is not valid. Either not hex or not " +
+                                                std::to_string(kAddressLength) + " bytes");
+                }
+
+                auto balance_str{item.value()["balance"].get<std::string>()};
+                Account account{0, intx::from_string<intx::uint256>(balance_str)};
+
+                // Make the account
+                account_changeset_table->put(block_number, *address_bytes);
+                plainstate_table->put(*address_bytes, account.encode_for_storage(true));
+
+                // Fills hash builder
+                auto hash{keccak256(*address_bytes)};
+                account_rlp[to_bytes32(hash.bytes)] = account.rlp(kEmptyRoot);
+            }
+
+            auto it{account_rlp.cbegin()};
+            trie::HashBuilder hb{full_view(it->first), it->second};
+            for (++it; it != account_rlp.cend(); ++it) {
+                hb.add(full_view(it->first), it->second);
+            }
+            root_hash = hb.root_hash();
```

Comment:
I think we should extract the allocation function into `silkworm/db` and add a test verifying that the state root matches what we expect for the main net, Rinkeby, etc. Presumably https://etherscan.io/block/0 has the correct state root.

Reply:
👍